### PR TITLE
[Bug] DFlash Training: removed causal constraint from noise mask for bidirectional attention within blocks

### PR DESCRIPTION
**Pull request by xiaomin-D**
## Motivation

This PR syncs with the latest mask-related fixes from upstream SpecForge.

Key upstream changes:
- `6272687`: Align mask_token_id logic with inference code
- `ca5455e`: Change dflash noise mask to bidirectional (remove causal constraint)

### Mask Logic Alignment
- Updated mask_token_id handling: auto-add `<|MASK|>` special token when tokenizer lacks mask_token_id
- Removed causal constraint from noise mask for bidirectional attention within blocks

## Related Issues

#412 

## Accuracy Test

WIP

## Benchmark & Profiling

WIP

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://sgl-fru7574.slack.com/archives/C09784E3EN6 to discuss your PR.

### CODE REVIEW

File: scripts/train_dflash.py [hunk 1]
```diff
@@ -310,19 +310,19 @@ def main():
...
         mask_token_id = args.mask_token_id
-    elif hasattr(tokenizer, "mask_token_id") and tokenizer.mask_token_id is not None:
+    elif tokenizer.mask_token_id is not None:
         mask_token_id = tokenizer.mask_token_id
-    elif hasattr(tokenizer, "pad_token_id") and tokenizer.pad_token_id is not None:
-        mask_token_id = tokenizer.pad_token_id
     else:
-        mask_token_id = tokenizer.eos_token_id
+        tokenizer.add_special_tokens({"mask_token": "<|MASK|>"})
+        mask_token_id = tokenizer.mask_token_id
```
[the remark this answers is not in the document]
Comment (xiaomin-D) [reply scripts/train_dflash.py:321]: This is not an issue in our case for the following reasons:
1. Qwen has reserved special token slots: Qwen models pre-reserve slots for special tokens in their vocabulary. The <|MASK|> token is assigned to one of these reserved slots (within the existing vocab_size), so no resizing is needed.
2. Draft model reuses target model's embeddings: The draft model shares the embedding layer from the target model, so there's no separate embedding to resize.

**Review by yubofredwang — APPROVED** [review]

**Comment by ggg-s** [comment]
hi @xiaomin-D,  does dflash require more GPU memory compared to eagle3?

**Comment by xiaomin-D** [comment]
> hi @xiaomin-D, does dflash require more GPU memory compared to eagle3?

yes，a little bit. Cause Dflash draft  model is bigger, but just a little bit~

**Comment by wengsnow** [comment]
Thanks for great work. I trained a Qwen3-8B draft model using the latest dflash code. While the training loss converged normally, the inference with dflash official code performance is poor. [code](https://github.com/z-lab/dflash?tab=readme-ov-file)
Results with sharegpt dataset:
1. sharegpt_train.jsonl : Acceptance length is 1.02.
image

2. sharegpt + ultrachat: Acceptance length is 2.06.
image

3. [official draft model](https://huggingface.co/z-lab/Qwen3-8B-DFlash-b16/tree/main)
image
The model seems to converge during training, but why is the acceptance length so low? Any advice on how to debug this? @xiaomin-D

**Comment by xiaomin-D** [comment]
> Thanks for great work. I trained a Qwen3-8B draft model using the latest dflash code. While the training loss converged normally, the inference with dflash official code performance is poor. [code](https://github.com/z-lab/dflash?tab=readme-ov-file) Results with sharegpt dataset:
> 
> 1. sharegpt_train.jsonl : Acceptance length is 1.02.
> 
> image
> 2. sharegpt + ultrachat: Acceptance length is 2.06.
> 
> image
> 3. [official draft model](https://huggingface.co/z-lab/Qwen3-8B-DFlash-b16/tree/main)
> 
> image The model seems to converge during training, but why is the acceptance length so low? Any advice on how to debug this? @xiaomin-D

Thanks for your experiments and for sharing the results!
Since ZLab hasn’t released the official training details yet, we currently have to reverse-engineer the setup from the inference code, so we can’t fully align with the official pipeline for now. We’ll follow up once their paper or code is public.

For the current situation:

**1.** Please try aligning the chat template. By default, the dflash model disables  </think>, while SpecForge enables it. A mismatch between training and inference chat templates can significantly affect acceptance length. The current SpecForge default keeps </think> mainly for better generality across different models and use cases.

**2.** On the loss side, different choices can matter. We use CE loss by default, but in our experiments, KL loss — as well as other loss variants — can sometimes give better results depending on the setup.

**3.** It may also help to further increase the dataset size, and make sure the dataset is regenerated , which tends to improve stability and acceptance length.

Although there is still a gap compared to the official model, with aligned chat templates and KL loss, we can reach around above 3 tokens on average acceptance length, for reference.

You can also try different training techniques, and any contributions would be appreciated.

**Comment by xiaomin-D** [comment]
@wengsnow

**Comment by wengsnow** [comment]
> > Thanks for great work. I trained a Qwen3-8B draft model using the latest dflash code. While the training loss converged normally, the inference with dflash official code performance is poor. [code](https://github.com/z-lab/dflash?tab=readme-ov-file) Results with sharegpt dataset:
> > 
> > 1. sharegpt_train.jsonl : Acceptance length is 1.02.
> > 
> > image
> > 2. sharegpt + ultrachat: Acceptance length is 2.06.
> > image
> > 3. [official draft model](https://huggingface.co/z-lab/Qwen3-8B-DFlash-b16/tree/main)
> > image The model seems to converge during training, but why is the acceptance length so low? Any advice on how to debug this? @xiaomin-D
> 
> Thanks for your experiments and for sharing the results! Since ZLab hasn’t released the official training details yet, we currently have to reverse-engineer the setup from the inference code, so we can’t fully align with the official pipeline for now. We’ll follow up once their paper or code is public.
> 
> For the current situation:
> 
> **1.** Please try aligning the chat template. By default, the dflash model disables , while SpecForge enables it. A mismatch between training and inference chat templates can significantly affect acceptance length. The current SpecForge default keeps mainly for better generality across different models and use cases.
> 
> **2.** On the loss side, different choices can matter. We use CE loss by default, but in our experiments, KL loss — as well as other loss variants — can sometimes give better results depending on the setup.
> 
> **3.** It may also help to further increase the dataset size, and make sure the dataset is regenerated , which tends to improve stability and acceptance length.
> 
> Although there is still a gap compared to the official model, with aligned chat templates and KL loss, we can reach around above 3 tokens on average acceptance length, for reference.
> 
> You can also try different training techniques, and any contributions would be appreciated.

Thank you very much for your helpful suggestions.
I realized that I indeed did not regenerate the dataset. I will regenerate it, retrain the model, and update the results as soon as possible.
I will also align the chat template and experiment with other loss functions.
